### PR TITLE
feat: Introduce `IPrivateKeyDecryptor` to allow using custom cryptography provider

### PR DESCRIPTION
**Pull request by lukaszsocha2**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -265,30 +265,44 @@ Javadocs are generated when `gradle javadoc` is run and can be found in
...
-The Box Java SDK uses libraries (`org.bouncycastle:bcpkix-jdk15on:1.57` and `org.bouncycastle:bcprov-jdk15on:1.57`) that are compatible with FIPS 140-2 validated cryptographic libraries (`org.bouncycastle:bc-fips:1.0.2.1`).
+To generate a Json Web Signature used for retrieving tokens in the JWT authentication method, the Box Java SDK decrypts an encrypted private key.
+For this purpose, Box Java SDK uses libraries (`org.bouncycastle:bcpkix-jdk15on:1.70` and `org.bouncycastle:bcprov-jdk15on:1.70`)
+that are NOT compatible with FIPS 140-2 validated cryptographic library (`org.bouncycastle:bc-fips`).
 
-### Vulnerabilities in Bouncycastle libraries
-In Box Java SDK we are using:
- - `org.bouncycastle:bcpkix-jdk15on:1.57`
- - `org.bouncycastle:bcprov-jdk15on:1.57`
+There are two ways of ensuring that decryption operation is FIPS compiant.
```
Comment (jimshowalter) [inline README.md:272]: FIPS compiant -> FIPS-compliant

File: src/main/java/com/box/sdk/JWTEncryptionPreferences.java [hunk 2]
```diff
@@ -81,4 +82,22 @@ public EncryptionAlgorithm getEncryptionAlgorithm() {
     public void setEncryptionAlgorithm(EncryptionAlgorithm encryptionAlgorithm) {
         this.encryptionAlgorithm = encryptionAlgorithm;
     }
+
+    /**
+     * Gets a decrytor used for decrypting the private key.
```
Comment (jimshowalter) [inline src/main/java/com/box/sdk/JWTEncryptionPreferences.java:87]: decrytor -> decryptor

File: src/test/java/com/box/sdk/BoxDeveloperEditionAPIConnectionTest.java [hunk 3]
```diff
@@ -159,6 +166,60 @@ public void retriesWithWhenJtiClaimIsDuplicated() {
         assertThat(api.getAccessToken(), is(accessToken));
     }
 
+    @Test
+    public void usesCustomDecryptorClassImplementation() throws NoSuchAlgorithmException, InvalidKeySpecException {
+        final String tokenPath = "/oauth2/token";
+        final String accessToken = "mNr1FrCvOeWiGnwLL0OcTL0Lux5jbyBa";
+        // This is freshly-generated private key, which is not used for a real Box account.
+        // It is safe to use in this unit test.
+        String decryprtedPrivateKey = "MIIEvgIBADANBgkqhkiG9w0BAQEFAASCBKgwggSkAgEAAoIBAQC5fehGf7yDfLTDewsBcPyJSIIH1IB"
```
Comment (jimshowalter) [inline src/test/java/com/box/sdk/BoxDeveloperEditionAPIConnectionTest.java:175]: decryprtedPrivateKey - decryptedPrivateKey

File: src/test/java/com/box/sdk/BoxDeveloperEditionAPIConnectionTest.java [hunk 4]
```diff
@@ -159,6 +166,60 @@ public void retriesWithWhenJtiClaimIsDuplicated() {
...
+        byte[] privateKeyBytes = Base64.decode(decryprtedPrivateKey);
+        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(privateKeyBytes);
+        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
+        PrivateKey privateKey = keyFactory.generatePrivate(keySpec);
+
+        IPrivateKeyDecryptor decryptorMock = mock(IPrivateKeyDecryptor.class);
+
+        when(decryptorMock.decryptPrivateKey(anyString(), eq("testkey"))).thenReturn(privateKey);
+        BoxDeveloperEditionAPIConnection api = this.getBoxDeveloperEditionAPIConnection(decryptorMock);
+
```
Comment (jimshowalter) [inline src/test/java/com/box/sdk/BoxDeveloperEditionAPIConnectionTest.java:202]: Remove extra blank line.

**Review by jimshowalter — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -265,30 +265,44 @@ Javadocs are generated when `gradle javadoc` is run and can be found in
...
 
-There are some moderate vulnerabilities reported against those versions:
+1. You can privide a custom implementation of the `IPrivateKeyDecryptor` interface, 
+which performs the decryption opertation using FIPS-certified library of your choice. The interface requires the
```
Comment (jimshowalter) [inline README.md:275]: opertation -> operation

**Review by jimshowalter — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (jimshowalter) [reply README.md:275]: privide -> provide

File: README.md [hunk 6]
```diff
@@ -265,30 +265,44 @@ Javadocs are generated when `gradle javadoc` is run and can be found in
...
+that are NOT compatible with FIPS 140-2 validated cryptographic library (`org.bouncycastle:bc-fips`).
 
-### Vulnerabilities in Bouncycastle libraries
-In Box Java SDK we are using:
- - `org.bouncycastle:bcpkix-jdk15on:1.57`
- - `org.bouncycastle:bcprov-jdk15on:1.57`
+There are two ways of ensuring that decryption operation is FIPS compiant.
 
-There are some moderate vulnerabilities reported against those versions:
+1. You can privide a custom implementation of the `IPrivateKeyDecryptor` interface,
```
Comment (jimshowalter) [inline README.md:274]: privide -> provide

**Review by jimshowalter — CHANGES_REQUESTED** [review]

File: src/test/java/com/box/sdk/BoxDeveloperEditionAPIConnectionTest.java [hunk 7]
```diff
@@ -159,6 +166,60 @@ public void retriesWithWhenJtiClaimIsDuplicated() {
         assertThat(api.getAccessToken(), is(accessToken));
     }
 
+    @Test
```
[the remark this answers is not in the document]
Comment (lukaszsocha2) [reply src/test/java/com/box/sdk/BoxDeveloperEditionAPIConnectionTest.java:169]: I don't use and real decryptor in this test, so it doesn't matter what private key I use. I do't have monaco-crypto dependency here

**Review by antusus — APPROVED** [review]
👍

**Pull request merged**